### ISSUE TITLE
Utility of an "index container" for PIConGPU kernels

### ISSUE THREAD

**Issue by n01r** [ComputationalRadiationPhysics/picongpu#919]
While refactoring the ionization framework I wondered if it was maybe useful to introduce an **"index container"** of some sort because the `typedef` section in kernels is relatively long and functors in these kernels have to do the same thing again. 

What I imagine is an object of some sort that gets initialized with each kernel call and each thread can deduce its own   
 - *global position*
 - *total position (inside moving window)*
 
both in units of cells and super cells, as `SIMDIM` vector and linear index
 - *position inside super cell* 
 - ...
  
with the same requirements.

The benefits: 
 - more consistency in namings for indices which improves readability and understandability of index operations
 - reduction of necessary code lines for `typedefs`

**Comment by n01r** [comment]
@psychocoderHPC, @ax3l 
what do you guys think?

**Comment by PrometheusPi** [comment]
I think, it's a great idea.

**Comment by ax3l** [comment]
@n01r I am missing that quite a bit, too. It's actually not a container but a high-level API. It won't work for everything, but abstracting the standard "which cell am I -> in super cell, on gpu, global" ([see naming conventions](https://github.com/ComputationalRadiationPhysics/picongpu/wiki/PIConGPU-domain-definitions)) and the question "which particle -> cell in super cell, cell on GPU, global cell"  would be the standard operations.

it's actually not much to add, but we should be consistent with out namings for domains.

**Comment by Heikman** [comment]
You can use cuSTL/container/IndexBuffer.hpp for this purpose. It is a virtual container of cell indices.

**Comment by Flamefire** [comment]
Currently the mappers can be queried for the super cell index:   
`const Space superCellIdx(mapper.getSuperCellIndex(Space(blockIdx)));`

But quite often you need the local cell index. So your write this over and over again:   
`const Space localCellIndex = (superCellIdx - mapper.getGuardingSuperCells()) * SuperCellSize::toRT() + threadIndex;`

So I propose adding a function to the mapper which gets the local cell index instead. Or at least something like supercell index without guards. Because in almost all getSuperCellIndex functions the guard cells are added to the index, just to be subtracted right after that.

So up to 3 functions can be added:

    - getLocalCellIndex(DataSpace<3> blockIdx, DataSpace<3> threadIdx)
    - getLocalCellIndex(DataSpace<3> blockIdx, uint32_t linearThreadIdx) (calls the above)
    - getLocalSuperCellIndex(DataSpace<3> blockIdx)

@ax3l Any ideas how to handle blockDims? My proposal adds support for linear and simDim-dimensional blocks as those are the most common ones. Only idea I have is to abstract that into "Super cell position" so that normally this is equal to threadIdx but you can modify it for cases where the block mapping is different.

**Comment by ax3l** [comment]
yes these optimizations are one aspect we should handle correctly in the interface (it should even encourage users to modify them).

I think we should be close to the concepts of the data structures we deal with when abstracting, so users just need to understand the concept (such as start/offset of a super cell, what a list of frame looks like, etc.) and can get the quantities we speak about naturally.

> Any ideas how to handle blockDims?

@psychocoderHPC will have the most educated opinion on that and we need to collect all current use cases and the ones we did not yet implement before designing an abstraction, because a bad design might obstruct manipulations we want to try out on changing hardware (e.g., causing complete kernel rewrites).

An alternative would be to simply not handle those exotic cases yet and to start with the most common ones, making many kernels more readable and a few stay for now quite low-level.

I changed the issue to a "feature request" in PMacc to reflect that we are evaluating concepts how to implement it.

**Comment by Flamefire** [comment]
Then the 3 functions I proposed would be a good start. The 3rd one is independent of any local mapping and is correct for all mappers (mappers map blocks to super cells)
The first function is the most common use case. The 2nd one covers the more exotic kernels, where linear blocks are started but map naturally to the 1st.

**Comment by ax3l** [comment]
> Then the 3 functions I proposed would be a good start. 

yep, think so, too.

**Comment by psychocoderHPC** [comment]
If someone implements the three funxtions above. Please do not add cell functionality to the mappers, this mixed two areas of abstractions in one class. 
We should also keep in mind that we call kernels wirh 3D mappers and one dimensional blocks size.
To avoid complex classes we should implement the new mapping method's outside of the current existing mapping classes.
@Flamefire Your second function can not call the first one, do to the missinf knowlage about the size if one block.

It would be nice if all functions supports compiletime abd runtime block sizes.

**Comment by Flamefire** [comment]
@psychocoderHPC So the functions may be added in another functor. I would that functor inherit or at least contain the mapper for easier access to all functions. All of them are mapping related...
About the 2nd->1st function call: We can assume (and document) that this function only applies for linearized thread ids which still refer to a super cell. Therefore the block size is always the super cell size. At the very least we can make this the default case and add support for different block sizes.
